src/Service/UserBookingCacheService.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+namespace App\Service;
+
+use App\Entity\Main\UserBooking;
+use App\Entity\Main\UserBookingCacheEntry;
+use App\Exception\MicrosoftGraphCommunicationException;
+use Doctrine\ORM\EntityManagerInterface;
+use Psr\Log\LoggerInterface;
+
+/**
+ * @see https://github.com/microsoftgraph/msgraph-sdk-php
+ * @see https://docs.microsoft.com/en-us/graph/use-the-api
+ */
+class UserBookingCacheService implements UserBookingCacheServiceInterface
+{
+    public function __construct(
+    private readonly MicrosoftGraphHelperService $graphHelperService,
+    private readonly EntityManagerInterface $entityManager,
+    private readonly MicrosoftGraphBookingService $microsoftGraphBookingService,
+    private readonly LoggerInterface $logger,
+  ) {
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \App\Exception\MicrosoftGraphCommunicationException
+     */
+    public function rebuildCache(): void
+    {
+        try {
+            $this->clearUserBookingCache();
+            $token = $this->graphHelperService->authenticateAsServiceAccount();
+            $nextResponse = $this->microsoftGraphBookingService->getAllFutureBookings($token);
+
+            // Loop over all pages of request.
+            while (isset($nextResponse)) {
+                $resultBody = $nextResponse->getBody();
+
+                // Loop over all elements on page.
+                foreach ($resultBody['value'] as $booking) {
+                    try {
+                        $userBooking = $this->microsoftGraphBookingService->getUserBookingFromApiData($booking);
+                        $this->entityManager->persist($this->createCacheEntity($userBooking));
+                    } catch (\Exception $e) {
+                        $this->logger->error($e->getMessage());
+                    }
+                }
+
+                // Determine next page.
+                if (isset($resultBody['@odata.nextLink'])) {
+                    $nextQuery = strstr($resultBody['@odata.nextLink'], '/me/events');
+                    $nextResponse = $this->graphHelperService->request($nextQuery, $token);
+                } else {
+                    $nextResponse = null;
+                }
+            }
+
+            // Add elements to db.
+            $this->entityManager->flush();
+        } catch (\Exception $e) {
+            throw new MicrosoftGraphCommunicationException($e->getMessage(), (int) $e->getCode());
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addCacheEntry(UserBooking $userBooking): void
+    {
+        $this->entityManager->persist($this->createCacheEntity($userBooking));
+        $this->entityManager->flush();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \Exception
+     */
+    public function changeCacheEntry(string $exchangeId, array $changes): void
```

Comment:
Make this method like createCacheEntity, with a line for each field in the entity.

```
!empty($changes['title]) && $entity->setTitle($changes['title]);
...
```

instead of the magic foreach below that might try to set a none-existing field